components/hab/src/command/pkg/download.rs

```diff
@@ -0,0 +1,439 @@
+        // Otherwise, it might make sense to take this fetch out of the verification code.
+        let nwr = artifact::artifact_signer(&artifact.path)?;
+        if SigKeyPair::get_public_key_path(&nwr, &self.path_for_keys()).is_err() {
+            ui.status(Status::Downloading,
+                      format!("Public key for signer {:?}", nwr))?;
+            self.fetch_origin_key(ui, &nwr, self.token)?;
+        }
+
+        if self.verify {
+            ui.status(Status::Verifying, artifact.ident()?)?;
+            artifact.verify(&self.path_for_keys())?;
+            debug!("Verified {} for {} signed by {}", ident, target, &nwr);
+        }
+        Ok(())
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    fn is_artifact_downloaded(&self, ident: &PackageIdent, target: PackageTarget) -> bool {
+        self.downloaded_artifact_path(ident, target).is_file()
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    /// Returns the path to the location this package would exist at in
+    /// the local package cache. It does not mean that the package is
+    /// actually *in* the package download directory, though.
+    fn downloaded_artifact_path(&self, ident: &PackageIdent, target: PackageTarget) -> PathBuf {
+        self.path_for_artifact()
+            .join(ident.archive_name_with_target(target).unwrap())
+    }
+
+    fn fetch_latest_package_in_channel_for(&self,
+                                           ident: &PackageIdent,
+                                           target: PackageTarget,
+                                           channel: &ChannelIdent,
+                                           token: Option<&str>)
+                                           -> Result<Package> {
+        let origin_package = self.api_client
+                                 .show_package_metadata((&ident, target), channel, token)?;
+        Ok(origin_package)
+    }
+
+    /// The cache_*_path functions in fs don't let you override a path base with Some(base)
+    /// So we have to build our own paths.
+    fn path_for_keys(&self) -> PathBuf { self.download_path.join("keys") }
+
+    fn path_for_artifact(&self) -> PathBuf { self.download_path.join("artifacts") }
+
+    /// Sanity check the download directory tree. The errors from the api around permissions are
+    /// opaque; this validates the directory in advance to help provide useful feedback.
+    fn verify_and_prepare_download_directory<T>(&self, ui: &mut T) -> Result<()>
+        where T: UIWriter
+    {
+        let system_paths = [self.download_path,
+                            &self.path_for_keys(),
+                            &self.path_for_artifact()];
+
+        ui.status(Status::Verifying,
+                  format!("Checking permissions and creating as needed the download directory \
+                           {:?}",
```

Comment:
Instead of using a debug formatter {:?}, we should be able to use a normal format {} with self.download_path.display()... same feedback for the other output below